### PR TITLE
Add types for IPv4, IPv6 and variants, including CIDR notation

### LINKED ISSUE #2: RFE: IP address
type ipv4address
type ipv6address
type ipaddress = Variant[ipv4address,ipv6address]

### PR DESCRIPTION
**Pull request by runejuhl**
Fixes #2

(I'm not opposed to adding tests, but I'd like a pointer to how to get there fast. I can see that stdlib uses the modulesync stuff, should this do the same?)

### CODE REVIEW

File: types/ipv4.pp [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# https://stackoverflow.com/a/190405/1000630, with minor modifications to syntax
+type Tea::IPv4 = Pattern[/((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.|$)){4}/]
```
Comment (ekohl) [inline types/ipv4.pp:2]: https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/compat/ipv4.pp

File: types/ipv6.pp [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Unsure where this came from, thank you Internet!
+type Tea::IPv6 = Pattern[/(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))/] # lint:ignore:140chars
```
Comment (ekohl) [inline types/ipv6.pp:2]: https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/compat/ipv6.pp

File: types/ip.pp [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# https://stackoverflow.com/a/190405/1000630, with minor modifications to syntax
+type Tea::IP = Variant[Tea::IPv4, Tea::IPv6]
```
Comment (ekohl) [inline types/ip.pp:2]: https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/compat/ip_address.pp

**Review by ekohl — COMMENTED** [review]
Would it make sense to add these CIDR types to stdlib instead?

For tests you can have a look at https://github.com/theforeman/puppet-dhcp/commit/7d31525491d362f090a31fad3bfe1802adcd59f4 but maybe rspec-puppet needs infrastructure to verify data types as well.

File: types/ip.pp (hunk 3 above)
Comment (runejuhl) [reply types/ip.pp:2]: Sure, adding them to stdlib would make sense. I talked a bit with @igalic about this at cfgmgmtcamp and I think we decided to start here, but having it in stdlib would be better. ...but in that case, why not move all types from tea into stdlib?

Comment (ekohl) [reply types/ip.pp:2]: That's entirely valid and this is partially already the case.

**Comment by runejuhl** [comment]
I can see that stdlib has added IP validation types (https://github.com/puppetlabs/puppetlabs-stdlib/commit/6dad21e292436ac5b7d8113e81e572a0efc66782), so I'll close this one.

**Comment by igalic** [comment]
\o/

**Pull request closed without merging**